thriftpool/doc_test.go

```diff
@@ -0,0 +1,97 @@
+package thriftpool_test
+
+import (
+	"context"
+	"time"
+
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/reddit/baseplate.go/thriftpool"
+)
+
+// In real code these should be coming from either config file or flags instead.
+const (
+	remoteAddr    = "host:port"
+	socketTimeout = time.Millisecond * 10
+
+	minConnections = 50
+	maxConnections = 100
+
+	clientTTL = time.Minute * 5
+)
+
+// BEGIN THRIFT GENERATED CODE SECTION
+//
+// In real code this section should be from thrift generated code instead,
+// but for this example we just define some placeholders here.
+
+type MyEndpointRequest struct{}
+
+type MyEndpointResponse struct{}
+
+type MyService interface {
+	MyEndpoint(ctx context.Context, req *MyEndpointRequest) (*MyEndpointResponse, error)
+}
+
+func NewMyServiceClientFactory(_ thrift.TTransport, _ thrift.TProtocolFactory) MyService {
+	// In real code this certainly won't return nil.
+	return nil
+}
+
+// END THRIFT GENERATED CODE SECTION
+
+type Client interface {
+	thriftpool.Client
+
+	MyService
+}
+
+type clientImpl struct {
+	MyService
+
+	*thriftpool.TTLClient
+}
+
+func newClient(addr string) (Client, error) {
+	trans, err := thrift.NewTSocketTimeout(addr, socketTimeout)
+	if err != nil {
+		return nil, err
+	}
+	err = trans.Open()
+	if err != nil {
+		return nil, err
+	}
+	protoFactory := thrift.NewTHeaderProtocolFactory()
+	client := NewMyServiceClientFactory(trans, protoFactory)
+	return &clientImpl{
+		MyService: client,
+		TTLClient: thriftpool.NewTTLClient(trans, clientTTL),
+	}, nil
+}
+
+func callEndpoint(ctx context.Context, pool thriftpool.Pool) (*MyEndpointResponse, error) {
+	c, err := pool.Get()
+	if err != nil {
+		return nil, err
+	}
+	defer pool.Release(c)
+
+	client := c.(Client)
+	return client.MyEndpoint(ctx, &MyEndpointRequest{})
+}
+
+// This example demonstrates a typical use case of thriftpool in microservice
+// code.
+func Example() {
+	pool, err := thriftpool.NewChannelPool(
+		minConnections,
+		maxConnections,
+		func() (thriftpool.Client, error) {
+			return newClient(remoteAddr)
+		},
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	callEndpoint(context.Background(), pool)
```

Comment:
Same here

Reply:
This is just an example calling the endpoint. In real code we certainly would also handle the response and errors :)